File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-1.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func kthSmallest(root *TreeNode, k int) int {
    var res int
	count := 0
	var dfs func(*TreeNode)
	dfs = func(root *TreeNode) {
		if root == nil {
			return
		} else if count == k {
			return
		}
		
		dfs(root.Left)
		count++
		if count == k  {
			res = root.Val
			return
		}
		dfs(root.Right)
		return
	}

	dfs(root)

	return res
}

File: Data Structures & Algorithms/binary-tree-diameter/submission-1.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @ param {TreeNode} root
     * @ param {[]int} res
     * @return number
     */
    dfs(root, res) {
        if (!root) {return 0}

        let left = this.dfs(root.left, res)
        let right = this.dfs(root.right, res)

        res[0] = Math.max(res[0], left + right)

        return 1 + Math.max(left, right)
    }
    /**
     * @param {TreeNode} root
     * @return {number}
     */
    diameterOfBinaryTree(root) {
        let res = [0]
        this.dfs(root, res)
        return res[0]
    }
}
